RFD 89: Merge Webapps into Teleport (#16413)
gravitational/teleport: rfd/0089-merge-webapps.md

---
authors: Jeff Pihach (jeff.pihach@goteleport.com)
state: draft
---

# RFD 89 - Merge Webapps into Teleport

## What

- Merge the [webapps](https://github.com/gravitational/webapps) repo into the
  [teleport](https://github.com/gravitational/teleport) repo.
- Merge the [webapps.e](https://github.com/gravitational/webapps.e) repo into
  the [teleport.e](https://github.com/gravitational/teleport.e) repo.

## Why

The version of the web UI and the Teleport are tightly coupled but the
repositories are separate which means that complexities arrise when integrating
the two projects. We have to separately clone and build the web UI and commit
the webasset changes into the teleport repo whenever we want to represent an
update. We also have to maintain multiple build systems and ops procedures for
the web UI. Moving the web UI repositories into their respective Teleport repo
will save us time and confusion.

> 

Reviewer: Might be worth adding that the webapps monorepo was developed at a time when we were supporting the UI for both Gravity and Teleport. Now that Gravity is no more, it makes more sense to combine them.

Author: done

## Success Criteria

Both [Teleport](https://github.com/gravitational/teleport) and
[Teleport.e](https://github.com/gravitational/teleport.e) repos contain their
respective UI counterparts and the build systems correctly build all projects.

## Process

Below outlines the process for each repository in the order that the steps need
to be taken. All repositories will be prepped for the final archival and merging
but the final switch over steps will not be taken until v11 has been released.

#### Webapps repository

[repository](https://github.com/gravitational/webapps)

- [ ] Triage issues.
  - [ ] Close those that are no longer relevant.
  - [ ] [Transfer remaining issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository) to the [Teleport repo](https://github.com/gravitational/teleport)
  - [ ] Add the [ui label](https://github.com/gravitational/teleport/labels/ui) to
        those transferred issues.
- [ ] Triage PRs.
  - [ ] Close those that are no longer relevant.

>   - [ ] Merge remaining PR's.

Reviewer: ```suggestion
  - [ ] Merge remaining PRs.
```

- update targets that use the `packages/webapps.e` submodule to point points to
  the correct version in the `teleport.e/webapps` folder.
- only require teleport build processes to run on teleport paths and the webapp
  ones to run on the webapp paths
-
- Archive the [Webapps repository](https://github.com/gravitational/webapps).

##### CI jobs

- **CodeQL / Analyze (go) (pull_request)**
- **Teleport-IntegrationTest (ci-account)**
- **Teleport-UnitTest (ci-account)**
- **Teleport-DocTest (ci-account)**
- **Teleport-Lint (ci-account)**
- **Teleport-OsCompatibility (ci-account)**
  - Disable for changes exclusively made in the `/webapp` path.
- **Assign / Auto Request Review**
  - Create a list of full stack engineers and have it pick from this list for
    changes made to the `/webapp` path.
- **Check / Checking reviewers**
  - Continue to require 2 reviewers for changes in the `webapp` projects.
- **Label / Label Pull Request (pull_request_target)**
  - Add label for UI to changes made in the `/webapp` path.
- **CodeQL / Analyze (javascript) (pull_request)**
  - Ensure that it's running for changes in `/webapp` path.
- **Code scanning results / CodeQL**
  - No changes

> - **webapps-build**

Reviewer: Would this job go away?

Author: Yes it would! Updated.

We should ensure that it can still be built though so we should create a new job to tackle that.

## What

- Merge the [webapps](https://github.com/gravitational/webapps) repo into the
  [teleport](https://github.com/gravitational/teleport) repo.
- Merge the [webapps.e](https://github.com/gravitational/webapps.e) repo into
  the [teleport.e](https://github.com/gravitational/teleport.e) repo.

## Why

The version of the web UI and the Teleport are tightly coupled but the
repositories were created separate to support both Gravity and Teleport. This
introduced complexities when integrating the two projects.

We have to separately clone and build the web UI and commit the webasset
changes into the teleport repo whenever we want to represent an update. We also
have to maintain multiple build systems and ops procedures for the web UI.

Now that Gravity is no longer a product we develop, moving the web UI
repositories into their respective Teleport repo will save us time and
confusion.

## Success Criteria

Both [Teleport](https://github.com/gravitational/teleport) and
[Teleport.e](https://github.com/gravitational/teleport.e) repos contain their
respective UI counterparts and the build systems correctly build all projects.

## Developer Experience

> ### Day to day

Reviewer 2: # How this will affect Connect

Generally not much as Connect is pretty self-contained. This is what comes to my mind on top of what Zac said.

## Default path to tsh

We'll have to update [the default path to tsh in dev mode](https://github.com/gravitational/webapps/blob/27c615b3ff6f317a85fac4aa28b8e73fa4aa0d28/packages/teleterm/src/mainProcess/runtimeSettings.ts#L18-L23). It's a simple change.

## Proto files

As mentioned by Zac, once webapps is merged into teleport, we want to streamline proto generation for Connect (#15955). Currently we update the protos in teleport, commit them there and then copy them over to webapps and commit the changes there as well. I suspect what we'll want to do is tell `buf` to generate Go proto files `lib/teleterm/api/protogen/golang` and the JS proto files inside `webapps/packages/teleterm/src/services/tshd/v1/`.

I'm not sure if #15955 needs to be implemented within this PR. It includes a lot of work already and we can live with copying the proto files for now.

Reviewer 2: Oh sorry, #15955 is about an even more obscure issue with protos in Connect. What I had in mind were the tshd protos generated with `make grpc-teleterm`, [as described in Connect's readme](https://github.com/gravitational/webapps/tree/27c615b3ff6f317a85fac4aa28b8e73fa4aa0d28/packages/teleterm#generating-tshd-grpc-protobuf-files).

We don't have to worry about #15955 at all, nor about Connect protos in general. We'll just keep copying them over, the location will just change slightly. Once the merge is completed we can tweak `make grpc-teleterm` so that copying is not necessary. I'll make an issue about it as we already have a bunch of issues related to Connect's proto files.

Reviewer: We should also add make targets for Teleport Connect that respect our existing OS/ARCH vars and correctly express the dependency on `build/tsh`.

> The Day to day development experience shouldn't change. The webapps project

Reviewer: Today we have the ability to run `DEBUG=1 teleport start ...` in order to load webassets from disk instead of the ones embedded in the binary.

Let's not lose this capability. You'll have to make a change to Teleport (probably before the migration) to update where it looks for webassets. Please play around and see if you can get this working, then update the "Actions" section at the end of this document.

Author: After playing with this a bit I think the best way to do this is to remove the `/webassets` submodule but keep the folder as the output location of the on-demand build. This will keep the functionality of `DEBUG=1` intact without any changes to `teleport` itself. I've updated the action below accordingly.

will continue to be self-contained and only require access to the nodejs and
yarn binaries. The locations of the enterprise assets will have to be changed
but that should remain transparent to the day to day activities.

To work on the enterprise version of the application the developer will have to
check out the enterprise version of Teleport which will bring with it the
appropriate webapp assets.

In the event that multiple developers are working on the same feature across

> the back and frontend work may need to be done on the same branch or the

Author: Actually that sentence was poorly written, I've rewritten it :)

developer may want to clone Teleport into another folder to build and update

> the backend binaries separate from the frontend.

Author: I added a reference to using a `git worktree` with the rewrite of this paragraph.

### Backports

To avoid having different proceses and build systems depending on the version

> of teleport, all branch versions (v9, v10, v11) will have their respective

Reviewer: ```suggestion
of teleport, all supported release branches (v9, v10, v11) will have their respective
versions of webapps migrated into the `teleport` repository.
```

> versions of the webapps repository.

Reviewer 2: Does this mean that the merge of webapps into teleport will be backported to v8, v10 and v11?

Author: Yes that's correct, well v9-11 :)

### Build process

The Teleport build process will need to be expanded to include nodejs and yarn
in order to be able to build the webapp assets and include into the final

> teleport binaries.

Author: This is mostly for those wanting to use `make full` to build a Teleport binary. As there will no longer be a `webassets` directory the UI will need to be built along with `make full` which requires node/yarn to do so.

> teleport binaries.

Reviewer: Let's be explicit here and mention that a compiled version of the webassets will no longer be committed to source control.

> teleport binaries.

Reviewer: ```suggestion
teleport binaries. A make target will be provided to perform a dockerized build for those who don't wish to install node and yarn on their workstation.
```

Reviewer 3: +1, both for not needing node/yarn and for the determinism this brings.

## Process

Below outlines the process for each repository in the order that the steps need
to be taken. All repositories will be prepped for the final archival and merging
but the final switch over steps will not be taken until v11 has been released.

> The git histories of each branch will be maintained while merging. [Merging repositories](https://stackoverflow.com/questions/13040958/merge-two-git-repositories-without-breaking-file-history)

Author: Agreed, the second answer there is the approach that I've been playing with.

  - [ ] [Transfer remaining issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository) to the [Teleport repo](https://github.com/gravitational/teleport)
  - [ ] Add the [ui label](https://github.com/gravitational/teleport/labels/ui) to
        those transferred issues.
- [ ] Triage PRs.
  - [ ] Close those that are no longer relevant.
  - [ ] Merge remaining PRs.
  - [ ] Apply the necessary backports.

### Webapps.e repository

[repository](https://github.com/gravitational/webapps.e)

#### Actions

- [ ] Triage issues.
  - [ ] Close those that are no longer relevant.
  - [ ] [Transfer remaining issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository) to the [Teleport.e repo](https://github.com/gravitational/teleport.e)
  - [ ] Add the [ui label](https://github.com/gravitational/teleport.e/labels/ui)
- [ ] Triage PRs.
  - [ ] Close those that are no longer relevant.
  - [ ] Merge remaining PR's.
  - [ ] Apply the necessary backports.

### Teleport repository

[repository](https://github.com/gravitational/teleport)

#### Actions

- [ ] Remove `/webassets` submodule

>   - This submodule is no longer required as the web UI will be built on demand.

Reviewer 4: Why not retain the `/webassets` submodule and have a script that updates it on every UI PR merge? The same behaviour we currently have (where every `webapps` merge updates the `/webassets` in `teleport`), except within the same repo.

Since the majority of devs working in the `teleport` repo don't make any UI changes, it seems like an unnecessary waste of time to have to build the UI on every Teleport build, this also means they wouldn't be required to have node and yarn installed on their machines. Another approach to mitigate this could be to have a specific `make` script that builds the Teleport binaries but doesn't build the UI. Any thoughts?

Reviewer: My 2 cents:

I don't like that the source code and the webassets submodule within the same commit/tag can be out of sync.

Dev's won't need to build the UI on _every_ teleport build, they'll only need to build it once to have a local copy of the assets and then any time they want to recompile UI changes.

- [ ] Clone the [Webapps repository](https://github.com/gravitational/webapps) into
      the Teleport root. [Maintaining their respective git histories](https://stackoverflow.com/questions/13040958/merge-two-git-repositories-without-breaking-file-history)
  - [ ] This will need to be done for every respective version branch (v9, v10, v11)
- [ ] Update targets that use the `packages/webapps.e` submodule to point points to
      the correct version in the `teleport.e/webapps` folder.
- [ ] Only require teleport build processes to run on teleport paths and the webapp
      ones to run on the webapp paths
-
- Archive the [Webapps repository](https://github.com/gravitational/webapps).

#### CI jobs

- **CodeQL / Analyze (go) (pull_request)**
- **Teleport-IntegrationTest (ci-account)**
- **Teleport-UnitTest (ci-account)**
- **Teleport-DocTest (ci-account)**

> - **Teleport-Lint (ci-account)**

Reviewer 3: Any changes in these? Or are we keeping webapps build/lint checks contained in the webapps- jobs (which I would actually recommend)?

Author: No changes to these, the webapps jobs will be moved over and run separately. This will allow us to skip out of jobs that don't need to run.... No need to run the `Teleport-Lint` job for example when changes were only made to the UI.

- **Teleport-OsCompatibility (ci-account)**

>   - Disable for changes exclusively made in the `/webapp` path.

Reviewer: Or just `/web`.

>   - Disable for changes exclusively made in the `/webapp` path.

Author: Updated the doc to use the `/web` folder name.

- **Assign / Auto Request Review**
  - Create a list of full stack engineers and have it pick from this list for
    changes made to the `/webapp` path.
- **Check / Checking reviewers**

>   - Continue to require 2 reviewers for changes in the `webapp` projects.

Reviewer: This unfortunately is not up for debate. We must enforce two reviewers for our SOC-2.

We got away with not enforcing 2 reviewers on webapps in the past because we required 2 approvals to update the submodule and incorporate the changes into teleport. Now that all the code will live in teleport, we'll be leveraging the same review process and requirements as any other teleport change.

>   - Continue to require 2 reviewers for changes in the `webapp` projects.

Reviewer 2: > This unfortunately is not up for debate. We must enforce two reviewers for our SOC-2.

Yes, I understand that. What I meant is not that we should maybe allow for just one reviewer but rather that in the past you had at least devs looking at your backend code and at least one fullstack looking at the frontend. Ending up with two backend devs looking at frontend changes or one backend and one fullstack dev feels just slightly worse in comparison.

>   - Continue to require 2 reviewers for changes in the `webapp` projects.

Author: > Ending up with two backend devs looking at frontend changes or one backend and one fullstack dev feels just slightly worse in comparison.

We are able to add additional requirements to the reviewer script so that changes to the `webapps` folder use a different list of developers. I can expand this line item to include that script change as well.

- **Label / Label Pull Request (pull_request_target)**
  - Add label for UI to changes made in the `/webapp` path.
- **CodeQL / Analyze (javascript) (pull_request)**
  - Ensure that it's running for changes in `/webapp` path.
- **Code scanning results / CodeQL**
  - No changes
- **webapps-build**
  - Instead of having this job build webassets it should ensure that it can be

>     built.

Reviewer 2: Isn't ensuring that webassets can be built the thing that this job does already?

Also, the `push-build-darwin-amd64` pipeline already runs `yarn install`, `yarn build-term` and `yarn package-term`. However, this only happens after a commit lands in teleport's master. I don't know if we do it for other platforms as well, we'd need to look into the dronegen stuff. The question is whether we want the push pipelines to still run on every commit to teleport.

The reason Connect is build during `push-build-darwin-amd64` is that it's one of the artifacts published on tag release. If there are problems with building Connect, [we want to discover them ASAP rather than wait for a tag release to break](https://gravitational.slack.com/archives/C0DF0TPMY/p1654618901893779?thread_ts=1654598627.161199&cid=C0DF0TPMY). Though TBH, at the moment the Connect build being done on push is somewhat different from the build being done on a tag release, but that's another story.

Author: What I meant to say was that it can just try to build the webassets not build and push. I'll update the copy.

Reviewer 2: Are we talking about the webapps-build CI job from `.cloudbuild/ci/build.yaml`? I don't think it pushes anything right now. 🤔

Reviewer 2: Ping; I don't think this jobs pushes webassets?

Reviewer 2: It's still not clear to me what kind of pushing this job does and will not do in the future. Am I mistaking it with some other job?

To work on the enterprise version of the application the developer will have to
check out the enterprise version of Teleport which will bring with it the
appropriate webapp assets.

Merging these repositories may cause some disruption when multiple developers
are working on the same feature across the back and frontend. Using a
[git worktree](https://git-scm.com/docs/git-worktree) to create a worktree of
the webapps folder is an elegant way of approaching this issue

### Backports

To avoid having different proceses and build systems depending on the version
of teleport, all supported release branches (v9, v10, v11) will have their
respective versions of the webapps repository.

### Build process

The Teleport build process will need to be expanded to include nodejs and yarn
in order to be able to build the webapp assets and include into the final
teleport binaries. We will no longer be committing a compiled version of the
webassets into the teleport repository, the UI will need to be compiled from
source on the initial build of teleport. A make target will be provided to
perform a dockerized build of the webassets for those who don't wish to install

> node and yarn on their workstation.

Reviewer 3: > Does anyone actually develop the teleport binary without embedding webassets?

I run debug sessions with `DEBUG=1 DEBUG_ASSETS_PATH=/path/to/webassets`. Super helpful to figure out flows that you are not familiar with.

> node and yarn on their workstation.

Reviewer 2: > I run debug sessions with `DEBUG=1 DEBUG_ASSETS_PATH=/path/to/webassets`. Super helpful to figure out flows that you are not familiar with.

Could you give an example of a scenario where you'd use this? It's not clear to me why I'd want to use those two env vars over recompiling teleport with embedded webassets or proxy updated Web UI to an existing cluster with `yarn start-teleport --target=https://example.com:3080/web`.

I suppose `DEBUG=1 DEBUG_ASSETS_PATH=…` replaces `yarn start-teleport` if you don't have Node.js tooling installed? Because then you can go to webapps, execute `make build-teleport` and then point `DEBUG_ASSETS_PATH` at assets that were just generated for you through Docker.

> node and yarn on their workstation.

Reviewer 3: That is not for Web UI development, but part of the invocation I use to debug the `teleport` binary itself for backend dev. I generally don't care about the webassets version, I just need the server to start.

> node and yarn on their workstation.

Author: I could potentially see a env variable `NO_UI=1 make full` or something to skip making the UI but I think that's out of scope for now.

[this comment](https://github.com/gravitational/teleport/pull/16413#discussion_r1001123524) outlines how this will work a bit. So one of the steps in the build process will create the built webassets and then the rest of the build will carry-on as it does now. I feel like this will be the quickest way to get this integrated and then we can adjust the process afterwards if we feel it needs adjusting.

> node and yarn on their workstation.

Reviewer 2: > That is not for Web UI development, but part of the invocation I use to debug the `teleport` binary itself for backend dev. I generally don't care about the webassets version, I just need the server to start.

@codingllama Ah, so you compile `teleport` without webassets and then use DEBUG & DEBUG_ASSETS_PATH just so that you can start the server?

> (…) one of the steps in the build process will create the built webassets and then the rest of the build will carry-on as it does now. I feel like this will be the quickest way to get this integrated and then we can adjust the process afterwards if we feel it needs adjusting.

@hatched Yeah, it'll definitely be the quickest way. And if other people use a similar workflow for backend dev to what Alan described then it might not be as disruptive.

## Process

Below outlines the process for each repository in the order that the steps need
to be taken. All repositories will be prepped for the final archival and merging
but the final switch over steps will not be taken until v11 has been released.

The git histories of each branch will be maintained while merging. [Merging repositories](https://stackoverflow.com/questions/13040958/merge-two-git-repositories-without-breaking-file-history)

### Webapps repository

[repository](https://github.com/gravitational/webapps)

#### Actions

- [ ] Triage issues.
  - [ ] Close those that are no longer relevant.
  - [ ] [Transfer remaining issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository) to the [Teleport repo](https://github.com/gravitational/teleport)
  - [ ] Add the [ui label](https://github.com/gravitational/teleport/labels/ui) to
        those transferred issues.
- [ ] Triage PRs.
  - [ ] Close those that are no longer relevant.
  - [ ] Merge remaining PRs.

>   - [ ] Apply the necessary backports.

Reviewer 2: How do we plan to orchestrate this? We cannot tell people to stop working until the switch is performed. But the way these points are stated here doesn't seem to imply that it'll be a seamless process.

Assuming the git history is kept, I think it could be possible to move any WIP branches by adding a remote in your local webapps clone which points to teleport, then push the branch to that remote, clone the branch in your local clone of teleport and rebase the branch on top of already migrated master of teleport.

Author: Ultimately the code will be 'moved over' one day so all work can continue on the existing branches until that time. The issue that this is trying to solve is that there are a number of open PR's that are stale so in having a cleaner list of open PR's will make the move simpler.

Reviewer 2: > Ultimately the code will be 'moved over' one day so all work can continue on the existing branches until that time. The issue that this is trying to solve is that there are a number of open PR's that are stale so in having a cleaner list of open PR's will make the move simpler.

Ah, thanks for the explanation, that makes sense. It looks like we were on the same page but I think this point about people being able to continue work on their regular branches wasn't expressed in the RFD so I wasn't sure how you planned around this.

  - [ ] Add the [ui label](https://github.com/gravitational/teleport.e/labels/ui)
- [ ] Triage PRs.
  - [ ] Close those that are no longer relevant.
  - [ ] Merge remaining PR's.
  - [ ] Apply the necessary backports.

### Teleport repository

[repository](https://github.com/gravitational/teleport)

#### Actions

- [ ] Remove `/webassets` submodule
  - This submodule is no longer required as the web UI will be built on-demand.
  - The folder will remain as the output location of the on-demand build but
    will not be committed.
- [ ] Clone the [Webapps repository](https://github.com/gravitational/webapps) into
      the Teleport root. [Maintaining their respective git histories](https://stackoverflow.com/questions/13040958/merge-two-git-repositories-without-breaking-file-history)
  - [ ] This will need to be done for every respective version branch (v9, v10, v11)
- [ ] Update targets that use the `packages/webapps.e` submodule to point points to
      the correct version in the `teleport.e/web` folder.
- [ ] Only require teleport build processes to run on teleport paths and the webapp
      ones to run on the webapp paths

> -

Reviewer 5: Somewhere here I'd add a point that Connect's build pipelines (in `teleport/dronegen`) needs to be updated - we won't have to clone `webapps` anymore.
> There's a number of // TODO in dronegen that will also be removed since the teleport connect code will already be present by default 

(from Zac's comment)

Author: done!

- [ ] Update Connect's build pipelines (in teleport/dronegen) as webapps will no longer need to be cloned.

#### CI jobs

- **CodeQL / Analyze (go) (pull_request)**
- **Teleport-IntegrationTest (ci-account)**
- **Teleport-UnitTest (ci-account)**
- **Teleport-DocTest (ci-account)**
- **Teleport-Lint (ci-account)**
- **Teleport-OsCompatibility (ci-account)**
  - Disable for changes exclusively made in the `/web` path.
- **Assign / Auto Request Review**
  - Create a list of full stack engineers and have it pick from this list for

>     changes made to the `/web` path.

Reviewer 2: How do we feel about sticking to the current process of manually picking reviewers?

It's not something we've discussed within the frontend team and as I mentioned [in my earlier comment](https://github.com/gravitational/teleport/pull/16413#discussion_r975315309), I feel it might be disruptive for the day-to-day work.
